sky/packages/sky/lib/src/gestures/tap.dart

```diff
@@ -2,37 +2,174 @@
+      isTrackingPointer = true;
+      router.addRoute(pointer, route);
+    }
+  }
+
+  void stopTrackingPointer(PointerRouter router, PointerRoute route) {
+    if (isTrackingPointer) {
+      isTrackingPointer = false;
+      router.removeRoute(pointer, route);
+    }
+  }
+
+  bool isWithinTolerance(PointerInputEvent event, double tolerance) {
+    ui.Offset offset = event.position - initialPosition;
+    return offset.distance <= tolerance;
+  }
+
+}
+
+/// TapGesture represents a full gesture resulting from a single tap
+/// sequence. Tap gestures are passive, meaning that they will not
+/// pre-empt any other arena member in play.
+class TapGesture extends TapTracker {
+
+  TapGesture({ this.gestureRecognizer, PointerInputEvent event })
+  : super(event: event) {
+      entry = GestureArena.instance.add(event.pointer, gestureRecognizer);
+      _wonArena = false;
+      _didTap = false;
+      startTimer(() => cancel());
+      startTrackingPointer(gestureRecognizer.router, handleEvent);
+    }
+
+  TapGestureRecognizer gestureRecognizer;
+
+  bool _wonArena;
+  bool _didTap;
 
+  void handleEvent(PointerInputEvent event) {
+    assert(event.pointer == pointer);
+    if (event.type == 'pointermove' && !isWithinTolerance(event, kTouchSlop)) {
+      cancel();
+    } else if (event.type == 'pointercancel') {
+      cancel();
+    } else if (event.type == 'pointerup') {
+      stopTimer();
+      stopTrackingPointer(gestureRecognizer.router, handleEvent);
+      _didTap = true;
+      _check();
+    }
+  }
+
+  void accept() {
+    _wonArena = true;
+    _check();
+  }
+
+  void reject() {
+    stopTimer();
+    stopTrackingPointer(gestureRecognizer.router, handleEvent);
+    gestureRecognizer._resolveTap(pointer, TapResolution.cancel);
+  }
+
+  void cancel() {
+    // If we won the arena already, then _entry is resolved, so resolving
+    // again is a no-op. But we still need to clean up our own state.
+    if (_wonArena)
+      reject();
+    else
+      entry.resolve(GestureDisposition.rejected);
+  }
+
+  void _check() {
+    if (_wonArena && _didTap)
+      gestureRecognizer._resolveTap(pointer, TapResolution.tap);
+  }
+
+}
+
+class TapGestureRecognizer extends DisposableArenaMember {
+  TapGestureRecognizer({ this.router, this.onTap, this.onTapDown, this.onTapCancel });
+
+  PointerRouter router;
   GestureTapCallback onTap;
   GestureTapCallback onTapDown;
   GestureTapCallback onTapCancel;
 
-  void handlePrimaryPointer(PointerInputEvent event) {
-    if (event.type == 'pointerdown') {
-      if (onTapDown != null)
-        onTapDown();
-    } else if (event.type == 'pointerup') {
-      resolve(GestureDisposition.accepted);
-      if (onTap != null)
-        onTap();
-    }
+  Map<int, TapGesture> _gestureMap = new Map<int, TapGesture>();
+
+  void addPointer(PointerInputEvent event) {
+    _gestureMap[event.pointer] = new TapGesture(
+      gestureRecognizer: this,
+      event: event
+    );
+    onTapDown?.call();
```

Comment:
Does this work? The analyzer is complaining about it (see travis log) and according to https://www.dartlang.org/articles/emulating-functions/ it's not something Functions implement, it's something you implement to emulate a Function.